### PR TITLE
fix: redact keras h5 lambda previews

### PR DESCRIPTION
**Pull request by mldangelo-oai**
## Summary

- Omit raw Keras H5 Lambda source and dict/list bytecode previews while retaining dangerous-pattern findings.
- Omit malformed Lambda configs and compiler error strings, which can echo attacker-controlled source lines containing secrets.
- Preserve current `main` hardening for mixed encoded functions, module references, bounded bytecode analysis, and false-positive-resistant pattern matching.
- Verify direct source, dict bytecode, legacy list bytecode, malformed syntax, and AST-valid/compiler-invalid source do not leak through JSON or SARIF.

## Validation

- `PROMPTFOO_DISABLE_TELEMETRY=1 uv run pytest tests/scanners/test_keras_h5_scanner.py -q` (`109 passed`)
- `PROMPTFOO_DISABLE_TELEMETRY=1 uv run pytest tests/scanners/test_keras_h5_scanner.py tests/scanners/test_keras_utils.py tests/scanners/test_keras_zip_scanner.py -q` (`359 passed`)
- `uv run ruff check modelaudit/ packages/modelaudit-picklescan/src packages/modelaudit-picklescan/tests tests/`
- `uv run ruff format --check modelaudit/ packages/modelaudit-picklescan/src packages/modelaudit-picklescan/tests tests/`
- `uv run mypy modelaudit/ packages/modelaudit-picklescan/src packages/modelaudit-picklescan/tests tests/` (`454 source files`)
- Full local xdist suite reached completion without a test failure before the known local parent-process shutdown stall; fresh GitHub CI is the authoritative full-matrix gate.
- `git diff --check`

## Scan Finding

- Fixes MA-DSS-C032 from the Codex Security scan.

### CODE REVIEW

**Comment by mldangelo-oai** [comment]
@codex review

File: modelaudit/scanners/keras_utils.py [hunk 1]
```diff
@@ -329,7 +330,7 @@ def check_lambda_dict_function(
                 "layer_class": "Lambda",
                 "dangerous_patterns": found_patterns,
                 "function_format": "dict_bytecode",
-                "code_preview": decoded_str[:200] + "..." if len(decoded_str) > 200 else decoded_str,
+                "code_preview": redact_evidence_string(decoded_str, max_chars=200),
```
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply modelaudit/scanners/keras_utils.py]: Fixed in 640bc88c. Dict-format Lambda bytecode no longer emits a raw code_preview because marshalled constants are opaque and cannot be safely redacted with source-text heuristics. The finding retains dangerous_patterns and now records code_preview_omitted=opaque_bytecode_may_contain_sensitive_constants. The regression serializes a real code object with a secret constant and verifies JSON and SARIF do not contain it.

**Review by ianw-oai — APPROVED** [review]

File: modelaudit/scanners/keras_h5_scanner.py [hunk 2]
```diff
@@ -952,7 +953,7 @@ def _check_lambda_layer(self, layer_config: dict[str, Any], result: ScanResult)
                         details={
                             "layer_class": "Lambda",
                             "code_analysis": risk_desc,
-                            "code_preview": function_str[:200] + "..." if len(function_str) > 200 else function_str,
+                            "code_preview": redact_evidence_string(function_str, max_chars=200),
```
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply modelaudit/scanners/keras_h5_scanner.py]: Closed in a2f9641b. The live head already omitted normal Lambda source previews entirely in 200d85c9, which covers compound AWS credential names in that path. This follow-up closes the adjacent malformed-source branch: suspicious invalid Lambda configs no longer persist raw layer_config values and instead record layer_config_omitted=lambda_config_may_contain_sensitive_literals. The end-to-end regression now includes aws_secret_access_key in malformed source and verifies the secret is absent from JSON and SARIF.

**Comment by mldangelo-oai** [comment]
Reviewed and updated against current `main` in `7f8b02df`.

Security QA found one additional output leak beyond the existing preview fixes: the AST-derived `code_analysis` string could serialize an attacker-controlled variadic Lambda argument name into JSON/SARIF. Dangerous H5 Lambda findings now omit that source-derived description and retain a stable `code_analysis_omitted` marker. The end-to-end regression includes a secret-bearing variadic identifier alongside source literals, marshalled bytecode constants, malformed configs, layer names, and safe/dangerous module references.

Validation:
- `tests/scanners/test_keras_h5_scanner.py`: 126 passed
- focused JSON/SARIF leak regression: passed
- scoped Ruff format/check: clean
- scoped mypy: clean
- all 2 review threads are resolved

The legacy helper test is gated off on this host's Python 3.12 reduced lane; its assertion was updated to the same contract directly covered by the passing H5 regression.

**Comment by mldangelo-oai** [comment]
Reviewed, synced with current `main`, and hardened beyond the original resolved comments.

Key QA/fixes:
- Replaced attacker-controlled Lambda layer names with deterministic scanner-owned ordinals so opaque, non-credential-shaped values cannot leak through CVE messages, locations, JSON, or SARIF.
- Omitted raw Lambda `module` and `function_name` evidence for both dangerous and safe-reference checks; findings retain status and explicit omission markers.
- Scoped malformed-source keyword detection to the actual function text with identifier boundaries. Benign metadata such as `evaluation_layer` or `retrieval quality` no longer creates a substring false positive, while malformed `eval`/`exec`/`compile`/`__import__` source remains detected.
- Expanded end-to-end JSON/SARIF regressions to cover opaque top-level/config layer names and both dangerous and safe module/function metadata.
- Merged the latest `main`; GitHub's prior conflict state resolved cleanly. Both existing review threads remain resolved.

Focused validation only (no full suite):
- `tests/scanners/test_keras_h5_scanner.py` plus `tests/utils/helpers/test_py_compile_improvements.py`: 127 passed, 8 expected Python 3.12 lane skips
- Scoped Ruff, format check, mypy, and `git diff --check`: clean

**Comment by mldangelo-oai** [comment]
QA follow-up on exact head `c7bda5a2`:

- Normalized all recognized qualified Lambda classes to the stable `Lambda` label in layer-count metadata and generic config findings, preventing artifact-controlled class paths from leaking into JSON/SARIF.
- Kept nested Lambda config context redacted so artifact-controlled config keys cannot leak through recursive finding paths.
- Made malformed Python builtin matching case-sensitive, avoiding false positives for non-executable names such as `EVAL`.
- Added JSON/SARIF regressions covering qualified class paths and nested config keys.

Validation:
- Focused Keras H5/compiler suite: 128 passed, 8 skipped
- Full suite: 10,419 passed, 815 skipped
- Full Ruff format/check: clean
- Full mypy: 454 files clean
- `uv lock --check`: clean

**Comment by mldangelo-oai** [comment]
Review follow-up pushed in 096bc04e. The H5 Lambda module-reference classifier no longer marks safe modules such as math or numpy critical solely because an unresolved attribute is named eval/system (including uppercase near-matches). Unqualified dangerous builtins and risky module roots such as os/subprocess remain flagged. Validation: 17 Keras utility tests passed, 7 focused classifier cases passed, qualified-Lambda evidence redaction probe passed; scoped Ruff, format, mypy, and diff checks are clean. H5-backed integration remains CI-only locally because h5py is unavailable.

**Comment by mldangelo-oai** [comment]
Audit complete on head `94a51afa`. In addition to the existing preview redaction, the combined fixes now: anonymize namespaced Lambda class counts and nested config keys; prevent fake `config.layer` metadata on Lambda from leaking through wrapper findings; fail closed on malformed scalar `function` metadata without echoing payloads; keep null placeholders benign; make malformed builtin matching case-sensitive; and avoid treating safe-module attributes like `math.eval` as dangerous solely by name. Validation on the exact pushed tree: focused union 14 passed; Keras surface 540 passed, 11 skipped before remote integration; Ruff format/check clean; mypy clean across 454 files; full canonical suite 10,309 passed, 858 skipped.

**Comment by mldangelo-oai** [comment]
Merged current main and resolved the Keras H5 overlap in 46e3c6aa. The resolution preserves main’s exact Lambda class allowlist, source/callback allowlists, auxiliary output_shape/mask scanning, nested callable handling, and fail-closed warnings while omitting source-derived code analysis, previews, validation errors, layer names, nested config paths, and module/function evidence. QA also separates generic near-matches such as math.eval and tensorflow.io.system (warning) from real callback bridges such as py_function, PyFunc, and load_op_library (critical). Targeted local validation: 25 passed; H5 integration files were skipped because h5py is unavailable locally; scoped Ruff, format, mypy, and git diff checks are clean. Both existing review threads remain resolved.

**Comment by mldangelo-oai** [comment]
Adversarial QA follow-up pushed at `390559bf` after integrating current `main`.

Fixed two remaining issues:
- updated the reduced-lane Lambda helper tests for synthetic, non-artifact-controlled layer names, closing the current Type Check failure
- contained Python syntax compilation inside a private temporary directory with an explicit bytecode path, preventing valid Lambda source and secret constants from persisting in orphaned `/tmp/__pycache__/*.pyc` files

Focused validation: 276 associated tests passed with 33 expected reduced-lane skips; scoped Ruff, mypy, and diff checks are clean. All review threads are resolved and the PR is mergeable/approved. Fresh CI is running; rotating without waiting.

File: modelaudit/utils/helpers/code_validation.py [hunk 3]
```diff
@@ -30,19 +30,15 @@
...
-            py_compile.compile(tmp_name, doraise=True)
-            return True, None
-        except py_compile.PyCompileError as e:
-            return False, str(e)
-        finally:
-            os.unlink(tmp_name)
+        with tempfile.TemporaryDirectory(prefix="modelaudit-python-validation-") as temp_dir:
+            source_path = Path(temp_dir) / "snippet.py"
+            bytecode_path = Path(temp_dir) / "snippet.pyc"
+            source_path.write_text(code, encoding="utf-8")
```
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply modelaudit/utils/helpers/code_validation.py]: Fixed in dfab222a. Lambda syntax validation no longer writes attacker-controlled source or bytecode to disk: it now performs ast.parse plus compile entirely in memory, preserving compiler-only semantic checks. Added reduced-lane coverage that fails if temporary-file APIs are used and verifies secret-bearing compiler errors are not echoed. Scoped helper/Lambda tests, Ruff, and mypy are clean; H5-backed tests await CI because h5py is unavailable locally.

**Comment by mldangelo-oai** [comment]
Critical review and remediation complete at `32f4f568` (fix commit `dfab222a`, followed by a current-`main` merge).

Review fixes:

- removed clear-text temporary storage of model-supplied Lambda source and bytecode; syntax validation now uses in-memory `ast.parse` plus `compile`
- retained compiler-only rejection of top-level `return`, `break`, `continue`, and `nonlocal` without echoing secret-bearing source in errors
- added reduced-Python-lane routing for the code-validation regressions
- added a guard that fails if syntax validation calls either temporary-file API
- stopped the generic Lambda configuration pass from case-folding Python identifiers, preventing benign `EVAL` and metadata near matches while keeping non-Lambda matching behavior unchanged

All review threads are addressed and resolved. The stale GitHub conflict state was cleared by merging current `main`; the changelog auto-merge retained all entries.

Local validation after the merge:

- runnable associated tests: 83 passed
- focused in-memory validation and Lambda helper tests: 58 passed
- H5-backed test modules: 2 skipped because `h5py` is unavailable locally
- Ruff format and lint: clean
- mypy on the six affected source/test files: clean
- diff whitespace check: clean

Fresh CI is running on the pushed head and will exercise the H5 end-to-end cases.

File: modelaudit/scanners/keras_h5_scanner.py [hunk 4]
```diff
@@ -1492,16 +1510,15 @@ def _parse_keras_version_components(version: str) -> tuple[tuple[int, int, int],
...
-                for token in re.split(r"[^0-9A-Za-z_]+", normalized_function_name)
-                if token.strip()
+                token.strip() for token in re.split(r"[^0-9A-Za-z_]+", normalized_function_name) if token.strip()
             }
-            if function_tokens & cls._DANGEROUS_LAMBDA_FUNCTION_NAMES:
+            if {token.lower() for token in function_tokens} & cls._ALWAYS_DANGEROUS_LAMBDA_FUNCTION_NAMES:
+                return True
+            if (
+                not isinstance(module_name, str) or not module_name.strip()
+            ) and function_tokens & cls._DANGEROUS_LAMBDA_FUNCTION_NAMES:
```
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply modelaudit/scanners/keras_h5_scanner.py:1535]: Fixed in e9595f3c. Legacy module-less callable tokens are now normalized to lowercase before matching, so `subprocess.Popen` remains a critical finding. Added an exact malicious regression plus the benign near-match `subprocess.PopenSafe`, which remains warning-only to guard against false positives. The associated Keras Lambda tests pass (311 passed, 8 reduced-lane skips), and scoped Ruff/mypy are clean.

**Comment by mldangelo-oai** [comment]
QA follow-up after syncing current main and addressing the remaining review thread:

- Fixed the mixed-case module-less callable false negative: `subprocess.Popen` is critical; benign near-match `subprocess.PopenSafe` remains warning-only.
- Made the deep-unary fail-closed regression portable across CPython parser depth differences (Python 3.10 may reject it during syntax validation; newer versions parse it and reject it from the allowlist).
- Associated Keras Lambda/helper tests: 311 passed, 8 reduced-lane skips.
- Re-ran the previously failing cache regression directly: 1 passed.
- Scoped Ruff format/check and mypy: clean.
- All review threads are addressed and resolved.

Pushed as e9595f3c; fresh CI is running.

**Comment by mldangelo-oai** [comment]
QA refresh after syncing current `main` at `58fbc9c6`:

- Resolved the single test import conflict by preserving both independent imports.
- Focused validation: `330 passed, 8 skipped` across the changed Keras H5/Lambda and shared code-validation test modules.
- Targeted Ruff check, Ruff format check, and mypy all pass on the changed source/test files.
- The prior Windows cache-identity failure was outside this PR's feature surface and is superseded by the newer base.

Updated head: `241ad87edda1f6c52e5ac1640fc002691574dd06`. Fresh CI is now running; moving on to the next review without waiting.

**Comment by mldangelo-oai** [comment]
Review follow-up pushed in `a6ded77d`.

The new module-qualified callable finding was valid: `io.open` could be downgraded to a generic warning. The fix uses a narrow dangerous module/function pair instead of treating every `open` or `system` name as critical. Added malicious and benign near-match coverage: `io.open` is critical; `io.open_safe` and `platform.system` remain warning-only.

Scoped validation: `tests/scanners/test_keras_h5_scanner.py` 250 passed; Ruff lint/format, mypy, and `git diff --check` are clean. Fresh CI is running.

File: modelaudit/scanners/keras_h5_scanner.py [hunk 5]
```diff
@@ -1498,15 +1516,14 @@ def _is_lambda_module_reference_dangerous(cls, module_name: Any, function_name:
...
-                if token.strip()
+                token.strip() for token in re.split(r"[^0-9A-Za-z_]+", normalized_function_name) if token.strip()
             }
-            if function_tokens & cls._DANGEROUS_LAMBDA_FUNCTION_NAMES:
+            if function_tokens & cls._ALWAYS_DANGEROUS_LAMBDA_FUNCTION_NAMES:
+                return True
+            if (
+                not isinstance(module_name, str) or not module_name.strip()
+            ) and function_tokens & cls._DANGEROUS_LAMBDA_FUNCTION_NAMES:
                 return True
```
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply modelaudit/scanners/keras_h5_scanner.py:1536]: Fixed in a6ded77d. Module-qualified danger is now handled with an exact `(module, function)` pair for `io.open`, preserving CRITICAL severity without promoting unrelated names globally. Regressions cover `io.open` as critical, `io.open_safe` as warning-only, and `platform.system` as warning-only to prevent false positives.
